sdk/metric/provider_test.go

```diff
@@ -174,3 +174,84 @@ func TestMeterProviderMixingOnRegisterErrors(t *testing.T) {
 		"Metrics produced for instrument collected by different MeterProvider",
 	)
 }
+
+func TestMeterProviderCardinalityLimit(t *testing.T) {
+	tests := []struct {
+		name        string
+		limit       int
+		expectedMax int // 0 => no limit
+	}{
+		{
+			name:        "no limit (default)",
+			limit:       -1, // special case: constructor without limit
+			expectedMax: 0,
+		},
+		{
+			name:        "no limit (limit=0)",
+			limit:       0,
+			expectedMax: 0,
+		},
+		{
+			name:        "no limit (negative)",
+			limit:       -5,
+			expectedMax: 0,
+		},
+		{
+			name:        "limit=5",
+			limit:       5,
+			expectedMax: 5,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			reader := NewManualReader()
+
+			var mp *MeterProvider
+			if tt.limit == -1 {
+				// default (no limit param)
+				mp = NewMeterProvider(WithReader(reader))
+			} else {
+				mp = NewMeterProvider(
+					WithReader(reader),
+					WithCardinalityLimit(tt.limit),
+				)
+			}
+
+			meter := mp.Meter("test-meter")
+
+			counter, err := meter.Int64Counter("metric")
+			if err != nil {
+				t.Fatalf("failed to create counter: %v", err)
+			}
+
+			for i := 0; i < 10; i++ {
+				counter.Add(context.Background(), 1,
+					api.WithAttributes(attribute.Int("key", i)))
+			}
+
+			var rm metricdata.ResourceMetrics
+			if err := reader.Collect(context.Background(), &rm); err != nil {
+				t.Fatalf("failed to collect metrics: %v", err)
+			}
+
+			data, ok := rm.ScopeMetrics[0].Metrics[0].Data.(metricdata.Sum[int64])
```

Comment:
This can panic.
Please use `require.Len` to make sure that `ScopeMetrics` and `Metrics` contain 1 element each.
Probably you can also use https://pkg.go.dev/github.com/stretchr/testify/require#IsType instead of type assertion.